File: helpdesk-pro/frontend/src/app/core/services/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { environment } from '../../../environments/environment';
import { ApiResponse, PageResponse, User, Role } from '../models';

@Injectable({ providedIn: 'root' })
export class UserService {
  private base = `${environment.apiUrl}/users`;

  constructor(private http: HttpClient) {}

  getUsers(role?: Role, page = 0, size = 20): Observable<PageResponse<User>> {
    let params = new HttpParams().set('page', page).set('size', size);
    if (role) params = params.set('role', role);
    return this.http.get<ApiResponse<PageResponse<User>>>(this.base, { params }).pipe(map(r => r.data));
  }

  getUser(id: string): Observable<User> {
    return this.http.get<ApiResponse<User>>(`${this.base}/${id}`).pipe(map(r => r.data));
  }

  updateUser(id: string, data: Partial<User>): Observable<User> {
    return this.http.put<ApiResponse<User>>(`${this.base}/${id}`, data).pipe(map(r => r.data));
  }

  deleteUser(id: string): Observable<void> {
    return this.http.delete<void>(`${this.base}/${id}`);
  }

  getMe(): Observable<User> {
    return this.http.get<ApiResponse<User>>(`${this.base}/me`).pipe(map(r => r.data));
  }

  updateProfile(data: { fullName?: string; currentPassword?: string; newPassword?: string }): Observable<User> {
    return this.http.put<ApiResponse<User>>(`${this.base}/me`, data).pipe(map(r => r.data));
  }
}

File: helpdesk-mobile/lib/core/services/user_service.dart
import '../api/api_client.dart';
import '../api/api_endpoints.dart';
import '../models/user_model.dart';

class UserService {
  final _api = ApiClient();

  Future<List<UserModel>> getUsers({String? role, int page = 0, int size = 100}) async {
    final params = <String, dynamic>{'page': page, 'size': size};
    if (role != null) params['role'] = role;
    final resp = await _api.get(ApiEndpoints.users, queryParams: params);
    final data = resp.data['data'];
    return (data['content'] as List).map((e) => UserModel.fromJson(e)).toList();
  }

  Future<UserModel> getMe() async {
    final resp = await _api.get(ApiEndpoints.me);
    return UserModel.fromJson(resp.data['data']);
  }

  Future<UserModel> updateProfile(Map<String, dynamic> data) async {
    final resp = await _api.put(ApiEndpoints.me, data: data);
    return UserModel.fromJson(resp.data['data']);
  }

  Future<UserModel> createUser(Map<String, dynamic> data) async {
    final resp = await _api.post(ApiEndpoints.users, data: data);
    return UserModel.fromJson(resp.data['data']);
  }

  Future<UserModel> updateUser(String id, Map<String, dynamic> data) async {
    final resp = await _api.put(ApiEndpoints.user(id), data: data);
    return UserModel.fromJson(resp.data['data']);
  }
}
